In [1]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical

import tensorflow as tf

%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

C:\ProgramData\Anaconda3\lib\site-packages\tensorflow\python\framework\dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
C:\ProgramData\Anaconda3\lib\site-packages\tensorflow\python\framework\dtypes.py:517: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
C:\ProgramData\Anaconda3\lib\site-packages\tensorflow\python\framework\dtypes.py:518: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
C:\ProgramData\Anaconda3\lib\site-packages\tensorflow\python\framework\dtypes.py:519: FutureWarning: Passi

In [2]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()

input_dim = 784
output_dim = nb_classes = 10

X_train = X_train.reshape(60000, input_dim)
X_test = X_test.reshape(10000, input_dim)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
X_train /= 255
X_test /= 255

In [3]:
Y_train = to_categorical(y_train, nb_classes)
Y_test = to_categorical(y_test, nb_classes)

## Part I

### I. Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use 8 as the mini-batch size.

In [4]:
model = Sequential()

model.add(Dense(128, input_shape=(784,), activation='relu'))
model.add(Dense(64, activation='relu'))
model.add(Dense(10, activation='softmax'))

Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor


In [5]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 128)               100480    
_________________________________________________________________
dense_1 (Dense)              (None, 64)                8256      
_________________________________________________________________
dense_2 (Dense)              (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________


In [6]:
model.compile(optimizer='sgd', loss='categorical_crossentropy', metrics=['accuracy'])

In [7]:
model.fit(X_train, Y_train, batch_size=8, epochs=20, verbose=1, validation_data=[X_test, Y_test])

Train on 60000 samples, validate on 10000 samples
Epoch 1/20
60000/60000 [==============================] - 16s 275us/sample - loss: 0.3680 - acc: 0.8957 - val_loss: 0.1994 - val_acc: 0.9423
Epoch 2/20
60000/60000 [==============================] - 10s 168us/sample - loss: 0.1756 - acc: 0.9483 - val_loss: 0.1395 - val_acc: 0.9578
Epoch 3/20
60000/60000 [==============================] - 9s 147us/sample - loss: 0.1269 - acc: 0.9628 - val_loss: 0.1132 - val_acc: 0.9662
Epoch 4/20
60000/60000 [==============================] - 9s 145us/sample - loss: 0.1004 - acc: 0.9704 - val_loss: 0.0960 - val_acc: 0.9711
Epoch 5/20
60000/60000 [==============================] - 9s 146us/sample - loss: 0.0820 - acc: 0.9761 - val_loss: 0.0900 - val_acc: 0.9730
Epoch 6/20
60000/60000 [==============================] - 9s 145us/sample - loss: 0.0691 - acc: 0.9795 - val_loss: 0.0858 - val_acc: 0.9723
Epoch 7/20
60000/60000 [==============================] - 9s 145us/sample - loss: 0.0585 - acc: 0.9827 - val

In [8]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test Score:{}'.format(score[0]))
print('Test Accuracy:{}'.format(score[1])) 

Test Score:0.07162760021513823
Test Accuracy:0.9787999987602234


### II. Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use 128 as the mini-batch size.

In [9]:
model = Sequential()

model.add(Dense(128, input_shape=(784,), activation='relu'))
model.add(Dense(64, activation='relu'))
model.add(Dense(10, activation='softmax'))

In [10]:
model.compile(optimizer='sgd', loss='categorical_crossentropy', metrics=['accuracy'])

In [11]:
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=1, validation_data=[X_test, Y_test])

Train on 60000 samples, validate on 10000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 22us/sample - loss: 1.2011 - acc: 0.6988 - val_loss: 0.5815 - val_acc: 0.8564
Epoch 2/20
60000/60000 [==============================] - 1s 17us/sample - loss: 0.4856 - acc: 0.8727 - val_loss: 0.3980 - val_acc: 0.8910
Epoch 3/20
60000/60000 [==============================] - 1s 17us/sample - loss: 0.3803 - acc: 0.8946 - val_loss: 0.3374 - val_acc: 0.9035
Epoch 4/20
60000/60000 [==============================] - 1s 18us/sample - loss: 0.3350 - acc: 0.9056 - val_loss: 0.3074 - val_acc: 0.9119
Epoch 5/20
60000/60000 [==============================] - 1s 17us/sample - loss: 0.3067 - acc: 0.9129 - val_loss: 0.2847 - val_acc: 0.9206
Epoch 6/20
60000/60000 [==============================] - 1s 17us/sample - loss: 0.2857 - acc: 0.9185 - val_loss: 0.2671 - val_acc: 0.9251
Epoch 7/20
60000/60000 [==============================] - 1s 17us/sample - loss: 0.2688 - acc: 0.9238 - val_loss: 0.

In [12]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test Score:{}'.format(score[0]))
print('Test Accuracy:{}'.format(score[1]))

Test Score:0.16621438756808637
Test Accuracy:0.9505000114440918


### III. Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use the full sample as the batch size.

In [13]:
model = Sequential()

model.add(Dense(128, input_shape=(784,), activation='relu'))
model.add(Dense(64, activation='relu'))
model.add(Dense(10, activation='softmax'))

In [14]:
model.compile(optimizer='sgd', loss='categorical_crossentropy', metrics=['accuracy'])

In [15]:
model.fit(X_train, Y_train, batch_size=X_train.shape[0], epochs=20, verbose=1, validation_data=[X_test, Y_test])

Train on 60000 samples, validate on 10000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 10us/sample - loss: 2.3470 - acc: 0.0978 - val_loss: 2.3423 - val_acc: 0.1010
Epoch 2/20
60000/60000 [==============================] - 0s 7us/sample - loss: 2.3410 - acc: 0.0989 - val_loss: 2.3363 - val_acc: 0.1015
Epoch 3/20
60000/60000 [==============================] - 0s 7us/sample - loss: 2.3352 - acc: 0.0999 - val_loss: 2.3305 - val_acc: 0.1019
Epoch 4/20
60000/60000 [==============================] - 0s 8us/sample - loss: 2.3295 - acc: 0.1010 - val_loss: 2.3248 - val_acc: 0.1032
Epoch 5/20
60000/60000 [==============================] - 0s 7us/sample - loss: 2.3239 - acc: 0.1022 - val_loss: 2.3192 - val_acc: 0.1048
Epoch 6/20
60000/60000 [==============================] - 0s 8us/sample - loss: 2.3185 - acc: 0.1039 - val_loss: 2.3138 - val_acc: 0.1061
Epoch 7/20
60000/60000 [==============================] - 0s 7us/sample - loss: 2.3132 - acc: 0.1052 - val_loss: 2.3085 -

In [16]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test Score:{}'.format(score[0]))
print('Test Accuracy:{}'.format(score[1]))

Test Score:2.2463470638275145
Test Accuracy:0.1396999955177307


#### Part I Summary.

Amongst the 3 models above, the first model with a batch size of 8 performed its best with a low score of 0.072 and an Accuracy of .98

## Part II

### I. Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use 0.01 as the learning rate.

In [17]:
model = Sequential()

model.add(Dense(128, input_shape=(784,), activation='relu'))
model.add(Dense(64, activation='relu'))
model.add(Dense(10, activation='softmax'))

In [18]:
model.summary()

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_9 (Dense)              (None, 128)               100480    
_________________________________________________________________
dense_10 (Dense)             (None, 64)                8256      
_________________________________________________________________
dense_11 (Dense)             (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________


In [19]:
model.compile(optimizer=tf.keras.optimizers.SGD(lr=.01), loss='categorical_crossentropy', metrics=['accuracy'])

In [20]:
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=1, validation_data=[X_test, Y_test])

Train on 60000 samples, validate on 10000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 20us/sample - loss: 1.3317 - acc: 0.6433 - val_loss: 0.6425 - val_acc: 0.8530
Epoch 2/20
60000/60000 [==============================] - 1s 19us/sample - loss: 0.5307 - acc: 0.8627 - val_loss: 0.4169 - val_acc: 0.8892
Epoch 3/20
60000/60000 [==============================] - 1s 18us/sample - loss: 0.4025 - acc: 0.8896 - val_loss: 0.3500 - val_acc: 0.9037
Epoch 4/20
60000/60000 [==============================] - 1s 17us/sample - loss: 0.3509 - acc: 0.9017 - val_loss: 0.3125 - val_acc: 0.9112
Epoch 5/20
60000/60000 [==============================] - 1s 17us/sample - loss: 0.3204 - acc: 0.9104 - val_loss: 0.2906 - val_acc: 0.9176
Epoch 6/20
60000/60000 [==============================] - 1s 17us/sample - loss: 0.2991 - acc: 0.9151 - val_loss: 0.2749 - val_acc: 0.9208
Epoch 7/20
60000/60000 [==============================] - 1s 17us/sample - loss: 0.2821 - acc: 0.9197 - val_loss: 0.

In [21]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test Score:{}'.format(score[0]))
print('Test Accuracy:{}'.format(score[1])) 

Test Score:0.172567222391814
Test Accuracy:0.949999988079071


### II. Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use 100 as the learning rate.

In [22]:
model = Sequential()

model.add(Dense(128, input_shape=(784,), activation='relu'))
model.add(Dense(64, activation='relu'))
model.add(Dense(10, activation='softmax'))

In [23]:
model.summary()

Model: "sequential_4"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_12 (Dense)             (None, 128)               100480    
_________________________________________________________________
dense_13 (Dense)             (None, 64)                8256      
_________________________________________________________________
dense_14 (Dense)             (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________


In [24]:
model.compile(optimizer=tf.keras.optimizers.SGD(lr=100), loss='categorical_crossentropy', metrics=['accuracy'])

In [25]:
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=1, validation_data=[X_test, Y_test])

Train on 60000 samples, validate on 10000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 23us/sample - loss: nan - acc: 0.0985 - val_loss: nan - val_acc: 0.0980
Epoch 2/20
60000/60000 [==============================] - 1s 21us/sample - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 3/20
60000/60000 [==============================] - 1s 18us/sample - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 4/20
60000/60000 [==============================] - 1s 19us/sample - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 5/20
60000/60000 [==============================] - 1s 18us/sample - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 6/20
60000/60000 [==============================] - 1s 19us/sample - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 7/20
60000/60000 [==============================] - 1s 18us/sample - loss: nan - acc: 0.0987 - val_loss: nan - val_acc: 0.0980
Epoch 8/20
60000/60

In [26]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test Score:{}'.format(score[0]))
print('Test Accuracy:{}'.format(score[1])) 

Test Score:nan
Test Accuracy:0.09799999743700027


### III. Implement a three-layer ANN model with 128, 64, and 10 neurons in the layers. Use 0.0000001 as the learning rate.

In [27]:
model = Sequential()

model.add(Dense(128, input_shape=(784,), activation='relu'))
model.add(Dense(64, activation='relu'))
model.add(Dense(10, activation='softmax'))

In [28]:
model.summary()

Model: "sequential_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_15 (Dense)             (None, 128)               100480    
_________________________________________________________________
dense_16 (Dense)             (None, 64)                8256      
_________________________________________________________________
dense_17 (Dense)             (None, 10)                650       
Total params: 109,386
Trainable params: 109,386
Non-trainable params: 0
_________________________________________________________________


In [29]:
model.compile(optimizer=tf.keras.optimizers.SGD(lr=0.0000001), loss='categorical_crossentropy', metrics=['accuracy'])

In [30]:
model.fit(X_train, Y_train, batch_size=128, epochs=20, verbose=1, validation_data=[X_test, Y_test])

Train on 60000 samples, validate on 10000 samples
Epoch 1/20
60000/60000 [==============================] - 1s 24us/sample - loss: 2.3688 - acc: 0.0741 - val_loss: 2.3689 - val_acc: 0.0736
Epoch 2/20
60000/60000 [==============================] - 1s 19us/sample - loss: 2.3688 - acc: 0.0741 - val_loss: 2.3689 - val_acc: 0.0737
Epoch 3/20
60000/60000 [==============================] - 1s 19us/sample - loss: 2.3688 - acc: 0.0741 - val_loss: 2.3689 - val_acc: 0.0738
Epoch 4/20
60000/60000 [==============================] - 1s 18us/sample - loss: 2.3688 - acc: 0.0741 - val_loss: 2.3689 - val_acc: 0.0738
Epoch 5/20
60000/60000 [==============================] - 1s 18us/sample - loss: 2.3688 - acc: 0.0741 - val_loss: 2.3689 - val_acc: 0.0738
Epoch 6/20
60000/60000 [==============================] - 1s 19us/sample - loss: 2.3688 - acc: 0.0741 - val_loss: 2.3689 - val_acc: 0.0738
Epoch 7/20
60000/60000 [==============================] - 1s 19us/sample - loss: 2.3688 - acc: 0.0742 - val_loss: 2.

In [31]:
score = model.evaluate(X_test, Y_test, verbose=0)
print('Test Score:{}'.format(score[0]))
print('Test Accuracy:{}'.format(score[1])) 

Test Score:2.3686777114868165
Test Accuracy:0.0738999992609024


#### Part II Summary.

Amongst the 3 models above, the first model with a batch size of 128  and a learning rate of .01 performed its best with a low score of 0.17 and an Accuracy of .95. The learning rate of .01 is the optimal learning rate among the 3 given. We cannot have a too large or too low learning rate when trying to reach the parameter with minimal lost. This model did not underestimate or overestimate.